tests: use test_interactive_test_util when possible

fjmolinas opened the pull request:
### Contribution description

In order to be able to run tests on as many boards as possible use `test_interactive_test_util` in all possible cases.

Many tests depend on resetting the board after the terminal has been opened to make sure the captured output is the start of the tests.

When `BOARD`'s can't provide this feature using `test_interactive_test_util` allows the application to wait for a `sync` from the test script to actually start.

Changes introduced:

- 0e1c79e: modify test runner to accept a `sync` option
- 28948d6: add shell commands for interactive sync
- faf8fa1: used interactive sync when possible. The module is added in `auto-init`. I did't split the commit since introducing it before the test changes would make most tests failed.

Can be splited from the PR:

- e5b1cd7: improve `tests/isr_yield_higher`

TODO in follow-up:

- reset before term
- not reset if `sync=True`:

```
    if sync is True:
        try:
            subprocess.check_output(('make', 'reset'), env=env,
                                    stderr=subprocess.PIPE)
        except subprocess.CalledProcessError:
            # make reset yields error on some boards even if successful
            pass
```


### Testing procedure

- Run all tests on multiple boards that support resetting, nothing should have changed:

<details><summary><b>native</b></summary>

```
python3 dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . native  --with-test-only
ERROR:native:Tests failed: 7
Failures during test:
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/lwip](tests/lwip/test.failed)
- [tests/netstats_l2](tests/netstats_l2/test.failed)
```
</details>

<details><summary><b>samr21-xpro</b></summary>

```
python3 dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . samr21-xpro  --with-test-only
Failures during test:
- [examples/suit_update](examples/suit_update/test.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)

```
</details>

<details><summary><b>iotlab-m3</b></summary>

```
ERROR:iotlab-m3:Tests failed: 7
Failures during test:
- [examples/suit_update](examples/suit_update/test.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)


```
</details>

- Green murdock to see I haven't missed any dependency

### Issues/PRs references

Part of #12448

fjmolinas commented:
With this PR test failed goes down from to 16 for ``z1`

<details><summary>**PR**:ERROR:z1:Tests failed: 76
</summary>

```
Failures during test:
- [tests/bench_runtime_coreapis](tests/bench_runtime_coreapis/test.failed)
- [tests/bench_sizeof_coretypes](tests/bench_sizeof_coretypes/test.failed)
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/build_system_cflags_spaces](tests/build_system_cflags_spaces/test.failed)
- [tests/buttons](tests/buttons/test.failed)
- [tests/cb_mux](tests/cb_mux/test.failed)
- [tests/cb_mux_bench](tests/cb_mux_bench/test.failed)
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/embunit](tests/embunit/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/evtimer_msg](tests/evtimer_msg/test.failed)
- [tests/float](tests/float/test.failed)
- [tests/fmt_print](tests/fmt_print/test.failed)
- [tests/gnrc_ipv6_nib](tests/gnrc_ipv6_nib/test.failed)
- [tests/gnrc_ipv6_nib_6ln](tests/gnrc_ipv6_nib_6ln/test.failed)
- [tests/gnrc_ndp](tests/gnrc_ndp/test.failed)
- [tests/gnrc_sock_ip](tests/gnrc_sock_ip/test.failed)
- [tests/gnrc_sock_udp](tests/gnrc_sock_udp/test.failed)
- [tests/heap_cmd](tests/heap_cmd/test.failed)
- [tests/irq](tests/irq/test.failed)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed)
- [tests/l2util](tests/l2util/test.failed)
- [tests/libc_newlib](tests/libc_newlib/test.failed)
- [tests/log_color](tests/log_color/test.failed)
- [tests/log_printfnoformat](tests/log_printfnoformat/test.failed)
- [tests/msg_avail](tests/msg_avail/test.failed)
- [tests/msg_send_receive](tests/msg_send_receive/test.failed)
- [tests/msg_try_receive](tests/msg_try_receive/test.failed)
- [tests/mutex_order](tests/mutex_order/test.failed)
- [tests/netdev_test](tests/netdev_test/test.failed)
- [tests/od](tests/od/test.failed)
- [tests/periph_gpio](tests/periph_gpio/test.failed)
- [tests/periph_timer](tests/periph_timer/test.failed)
- [tests/pipe](tests/pipe/test.failed)
- [tests/pkg_c25519](tests/pkg_c25519/test.failed)
- [tests/pkg_cayenne-lpp](tests/pkg_cayenne-lpp/test.failed)
- [tests/pkg_heatshrink](tests/pkg_heatshrink/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed)
- [tests/pkg_nanocbor](tests/pkg_nanocbor/test.failed)
- [tests/pkg_tiny-asn1](tests/pkg_tiny-asn1/test.failed)
- [tests/pkg_tweetnacl](tests/pkg_tweetnacl/test.failed)
- [tests/pkg_ucglib](tests/pkg_ucglib/test.failed)
- [tests/pkg_umorse](tests/pkg_umorse/test.failed)
- [tests/posix_semaphore](tests/posix_semaphore/test.failed)
- [tests/pthread](tests/pthread/test.failed)
- [tests/pthread_barrier](tests/pthread_barrier/test.failed)
- [tests/pthread_cleanup](tests/pthread_cleanup/test.failed)
- [tests/pthread_condition_variable](tests/pthread_condition_variable/test.failed)
- [tests/pthread_cooperation](tests/pthread_cooperation/test.failed)
- [tests/pthread_flood](tests/pthread_flood/test.failed)
- [tests/pthread_rwlock](tests/pthread_rwlock/test.failed)
- [tests/pthread_tls](tests/pthread_tls/test.failed)
- [tests/riotboot_hdr](tests/riotboot_hdr/test.failed)
- [tests/rmutex](tests/rmutex/test.failed)
- [tests/sched_testing](tests/sched_testing/test.failed)
- [tests/shell](tests/shell/test.failed)
- [tests/test_tools](tests/test_tools/test.failed)
- [tests/thread_basic](tests/thread_basic/test.failed)
- [tests/thread_cooperation](tests/thread_cooperation/test.failed)
- [tests/thread_exit](tests/thread_exit/test.failed)
- [tests/thread_flags](tests/thread_flags/test.failed)
- [tests/thread_flags_xtimer](tests/thread_flags_xtimer/test.failed)
- [tests/thread_flood](tests/thread_flood/test.failed)
- [tests/thread_msg](tests/thread_msg/test.failed)
- [tests/thread_msg_block_w_queue](tests/thread_msg_block_w_queue/test.failed)
- [tests/thread_msg_block_wo_queue](tests/thread_msg_block_wo_queue/test.failed)
- [tests/thread_msg_seq](tests/thread_msg_seq/test.failed)
- [tests/thread_race](tests/thread_race/test.failed)
- [tests/trickle](tests/trickle/test.failed)
- [tests/xtimer_hang](tests/xtimer_hang/test.failed)
- [tests/xtimer_msg_receive_timeout](tests/xtimer_msg_receive_timeout/test.failed)
- [tests/xtimer_now64_continuity](tests/xtimer_now64_continuity/test.failed)
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/test.failed)
- [tests/xtimer_remove](tests/xtimer_remove/test.failed)
- [tests/xtimer_reset](tests/xtimer_reset/test.failed)
- [tests/xtimer_usleep](tests/xtimer_usleep/test.failed)

```
</details>


<details><summary>**PR**:ERROR:z1:Tests failed: 16
</summary>

```
Failures during compilation:
- [tests/gnrc_ndp](tests/gnrc_ndp/compilation.failed)

Failures during test:
- [tests/bench_runtime_coreapis](tests/bench_runtime_coreapis/test.failed)
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/evtimer_msg](tests/evtimer_msg/test.failed)
- [tests/periph_gpio](tests/periph_gpio/test.failed)
- [tests/pkg_c25519](tests/pkg_c25519/test.failed)
- [tests/pkg_heatshrink](tests/pkg_heatshrink/test.failed)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed)
- [tests/pkg_tiny-asn1](tests/pkg_tiny-asn1/test.failed)
- [tests/pkg_tweetnacl](tests/pkg_tweetnacl/test.failed)
- [tests/pthread_tls](tests/pthread_tls/test.failed)
- [tests/shell](tests/shell/test.failed)
- [tests/thread_flags_xtimer](tests/thread_flags_xtimer/test.failed)
```
</details>

Te compilation issue in tests/gnrc_ndp is because of fw size.

fjmolinas commented:
I'm going to trigger murdock to see which applications don't compile because of size anymore.

keestux commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -22,10 +22,12 @@
 
 
-def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
+def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False, sync=False):
```
Have you considered to make sync default to True? Wouldn't that reduce the number of files you have to change for this PR?

fjmolinas replied:
Sure, I didn't initially because I wasn't sure if the PR would need to be split and introduce changes slowly, in which case a default `false` was better.

I can remove them at the end, for now I rather keep it help me see easily which tests I have changed and which not :)

fjmolinas replied:
> I can remove them at the end, for now I rather keep it help me see easily which tests I have changed and which not :)

By the end I mean after squashing :)

MrKevinWeiss replied:
Maybe it is time now, if all tests are not regressing, please squash and update the with default `sync=True`

fjmolinas replied:
I'll test the squash version then push.

keestux commented on tests/isr_yield_higher/main.c:
```diff
@@ -44,24 +45,26 @@ static void _cb(void *arg)
+            second_thread, &start_time, "nr2");
+
+
+    while(xtimer_now_usec() < (TEST_TIME + start_time)) {}
```
Why not insert a space after the `while` now that this line is changed?

fjmolinas replied:
Addressed in #12539

keestux commented on tests/README.md:
```diff
@@ -7,6 +7,27 @@ allow basic functionality to be verified as well as provide an example of
+The goal is to be able to run all tests in a sequential way for as many targets
+as possible.
+
+As some board can't be rested without a manual trigger tests should be implemented
```
rested => reset?

keestux commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -22,10 +22,12 @@
     child = setup_child(timeout, env=os.environ,
                         logfile=sys.stdout if echo else None)
     try:
+        if sync is True:
```
Why not simply?
```
    if sync:
```

keestux commented on sys/shell/commands/shell_commands.c:
```diff
@@ -267,6 +272,10 @@ const shell_command_t _shell_command_list[] = {
+#endif
+#ifdef MODULE_TEST_UTILS_INTERACTIVE_SYNC
+    { "r", "Test sync, Ready query", _r },
+    { "s", "Test sync, Start test tiger", _s },
```
tiger => trigger?

fjmolinas commented:
rebased

fjmolinas commented:
rebased

MrKevinWeiss commented:
I can make this a priority if you have time to work on it (I think we need to move fast due to moving target rebasing).

My only concern with this is that it changes the behaviour of the tests when you just want to term in and read values. if there is a way to include everything only when `make test` is called and not when `make term` is.

Also I think it would be good to evaluate this on the arduino-mega2560.  I have had some problems with getting stuck in the bootloader mode when trying to poll if the terminal is ready (I believe you need a 2.6 second timeout after opening a serial connection otherwise it could take up to 10 seconds or so to go out of bootloader mode, this 10 seconds resets whenever you try to talk to it).  I don't see this problem on a cheap offbrand china version but on the standard ones this is how the bootloader seems to behave.

fjmolinas commented:
> I can make this a priority if you have time to work on it (I think we need to move fast due to moving target rebasing).

I was on holiday but I'm back and can get back to working on this.

> Also I think it would be good to evaluate this on the arduino-mega2560. I have had some problems with getting stuck in the bootloader mode when trying to poll if the terminal is ready (I believe you need a 2.6 second timeout after opening a serial connection otherwise it could take up to 10 seconds or so to go out of bootloader mode, this 10 seconds resets whenever you try to talk to it). I don't see this problem on a cheap offbrand china version but on the standard ones this is how the bootloader seems to behave.

I will test and see what happens.

> My only concern with this is that it changes the behaviour of the tests when you just want to term in and read values. if there is a way to include everything only when make test is called and not when make term is.

I'll look into this

fjmolinas commented:
rebased

MrKevinWeiss commented:
OK, I will provide attention to this as well, ping if needed!

fjmolinas commented:
@MrKevinWeiss results when running all tests on mastet and this PR for `atmega-2560`:

<details><summary><b>PR</b></summary>

```
Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed)
- [tests/libfixmath](tests/libfixmath/test.failed)
- [tests/periph_cpuid](tests/periph_cpuid/test.failed)
- [tests/pipe](tests/pipe/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_libb2](tests/pkg_libb2/test.failed)
- [tests/pkg_micro-ecc](tests/pkg_micro-ecc/test.failed)
- [tests/pkg_nanopb](tests/pkg_nanopb/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/pthread_barrier](tests/pthread_barrier/test.failed)
- [tests/trickle](tests/trickle/test.failed)
```
</details>

<details><summary><b>master</b></summary>

```
Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed)
- [tests/libfixmath](tests/libfixmath/test.failed)
- [tests/periph_wdt](tests/periph_wdt/test.failed)
- [tests/pipe](tests/pipe/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_libb2](tests/pkg_libb2/test.failed)
- [tests/pkg_micro-ecc](tests/pkg_micro-ecc/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/trickle](tests/trickle/test.failed)
```
</details>


I'll look into the diff.

MrKevinWeiss commented:
for the RF tests I added a CONNECT_WAIT env variable.  This stopped anything from being sent for n after connecting.

MrKevinWeiss commented:
Tested with `RIOT_TERMINAL` socat, picocom, pyterm.  All seem good.

fjmolinas commented:
> - [tests/periph_cpuid](tests/periph_cpuid/test.failed)

This was because it was a new test, fixed.

> - [tests/pkg_nanopb](tests/pkg_nanopb/test.failed)

I was missing the package requirements, fixed.

> - [tests/pthread_barrier](tests/pthread_barrier/test.failed)

This test Timeouts, I increased the timeout.

No diff with respect to `arduino-mega2560`

There was `tests/cpu_efm32_features/` that was added in the meantime which I had not included, so I added it

<details><summary>BOARD=slstk3402a make -C tests/cpu_efm32_features/ flash test
</summary>

```
Reset delay: 0 ms
Reset type NORMAL: Resets core & peripherals via SYSRESETREQ & VECTRESET bit.
Reset: Halt core after reset via DEMCR.VC_CORERESET.
Reset: Reset device via AIRCR.SYSRESETREQ.



Script processing completed.

r
/home/francisco/workspace/RIOT/dist/tools/pyterm/pyterm -p "/dev/riot/tty-slstk3402a" -b "115200" --noprefix --no-repeat-command-on-empty-line
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/riot/tty-slstk3402a
Welcome to pyterm!
Type '/exit' to exit.
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2020.01-devel-635-g50ce5-pr_apply_interactive_sync_tests)
Board booted, with some EFM32 features enabled or disabled.

make: Leaving directory '/home/francisco/workspace/RIOT/tests/cpu_efm32_features'
```
</details>

fjmolinas commented:
> My only concern with this is that it changes the behaviour of the tests when you just want to term in and read values. if there is a way to include everything only when make test is called and not when make term is.

I do not think there is a way around this, it would mean changing the compiled fw. But I understand you concern, I'll right to the developer mainling list just to make sure everyone is OK.

MrKevinWeiss commented:
Sure, we have needed something like this for a while.  I think it is an improvement overall, once you let everyone know we can continue.

fjmolinas commented:
I re-ran the tests, not tests are failing that aren't already failing in master.

<details><summary></summary>

```
#### arduino-mega2560/failuresummary.md

Failures during test:
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/isr_yield_higher](tests/isr_yield_higher/test.failed)
- [tests/libfixmath](tests/libfixmath/test.failed)
- [tests/pipe](tests/pipe/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)
- [tests/pkg_jsmn](tests/pkg_jsmn/test.failed)
- [tests/pkg_libb2](tests/pkg_libb2/test.failed)
- [tests/pkg_micro-ecc](tests/pkg_micro-ecc/test.failed)
- [tests/ps_schedstatistics](tests/ps_schedstatistics/test.failed)
- [tests/shell](tests/shell/test.failed)
- [tests/trickle](tests/trickle/test.failed)

#### frdm-k64f/failuresummary.md

Failures during test:
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/periph_timer](tests/periph_timer/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)

#### iotlab-m3/failuresummary.md

Failures during test:
- [examples/suit_update](examples/suit_update/test.failed)
- [tests/driver_hd44780](tests/driver_hd44780/test.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)

#### nucleo-l152re/failuresummary.md

Failures during test:
- [examples/suit_update](examples/suit_update/test.failed)
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/gnrc_ipv6_ext](tests/gnrc_ipv6_ext/test.failed)
- [tests/gnrc_ipv6_ext_frag](tests/gnrc_ipv6_ext_frag/test.failed)
- [tests/gnrc_rpl_srh](tests/gnrc_rpl_srh/test.failed)
- [tests/gnrc_sock_dns](tests/gnrc_sock_dns/test.failed)
- [tests/gnrc_tcp](tests/gnrc_tcp/test.failed)
- [tests/pkg_fatfs_vfs](tests/pkg_fatfs_vfs/test.failed)

#### z1/failuresummary.md

Failures during compilation:
- [tests/gnrc_ipv6_nib](tests/gnrc_ipv6_nib/compilation.failed)
- [tests/gnrc_ipv6_nib_6ln](tests/gnrc_ipv6_nib_6ln/compilation.failed)
- [tests/gnrc_ndp](tests/gnrc_ndp/compilation.failed)
- [tests/gnrc_sock_ip](tests/gnrc_sock_ip/compilation.failed)
- [tests/gnrc_sock_neterr](tests/gnrc_sock_neterr/compilation.failed)
- [tests/gnrc_sock_udp](tests/gnrc_sock_udp/compilation.failed)

Failures during test:
- [tests/bench_runtime_coreapis](tests/bench_runtime_coreapis/test.failed)
- [tests/bitarithm_timings](tests/bitarithm_timings/test.failed)
- [tests/driver_grove_ledbar](tests/driver_grove_ledbar/test.failed)
- [tests/driver_hd44780](tests/driver_hd44780/test.failed)
- [tests/driver_my9221](tests/driver_my9221/test.failed)
- [tests/event_wait_timeout](tests/event_wait_timeout/test.failed)
- [tests/evtimer_msg](tests/evtimer_msg/test.failed)
- [tests/periph_gpio](tests/periph_gpio/test.failed)
- [tests/pkg_c25519](tests/pkg_c25519/test.failed)
- [tests/pkg_heatshrink](tests/pkg_heatshrink/test.failed)
- [tests/pkg_lora-serialization](tests/pkg_lora-serialization/test.failed)
- [tests/pkg_tiny-asn1](tests/pkg_tiny-asn1/test.failed)
- [tests/pkg_tweetnacl](tests/pkg_tweetnacl/test.failed)
- [tests/pthread_tls](tests/pthread_tls/test.failed)
- [tests/shell](tests/shell/test.failed)
- [tests/thread_flags_xtimer](tests/thread_flags_xtimer/test.failed)
```

</details>

MrKevinWeiss commented:
rerunning tests as well.  @aabadie @keestux if you have anything to say now is the time!

fjmolinas commented:
I pushed removing an unrelated change that was causing a PEP8 error.

aabadie commented on tests/README.md:
```diff
@@ -7,6 +7,27 @@ allow basic functionality to be verified as well as provide an example of
+The module for the first option is `test_utils_interactive_sync` and is set as a
+default module in `Makefile.tests_common`. It can be disabled by setting in the
+application makefile `DISABLE_MODULE += test_utils_interactive_sync`. Then in
+the automatic python script `sync=False` should be set as a argument of `run(..)`
```
Would it make sense to export an environment variable `TEST_USE_INTERACTIVE_SYNC` when `test_utils_interactive_sync` is in `DISABLE_MODULE` and set `sync=False` in this case:

```mk
ifneq (,$(filter test_utils_interactive_sync,$(DISABLE_MODULE))
  export TEST_USE_INTERACTIVE_SYNC=1
endif
```

and in `testrunner/__init__.py`:

```py
sync = int(os.getenv('TEST_USE_INTERACTIVE_SYNC', 0)) == 1
```

This would basically allow to not have to change the scripts in the tests. There are maybe things that I'm missing though.

fjmolinas replied:
I don't see why editing the files is an issue to be honest. But sure you could do it that way, although in that case its simple to do it how @cladmi had done it before, some kind of variant of:

```
    modules = set(os.environ.get('USEMODULE', '').split(' '))
    try:
        if 'test_utils_interactive_sync' in modules:
            test_utils_interactive_sync(child)
        testfunc(child)
```

fjmolinas commented on tests/pthread_barrier/tests/01-run.py:
```diff
@@ -16,4 +16,4 @@ def testfunc(child):
 
 if __name__ == "__main__":
-    sys.exit(run(testfunc))
+    sys.exit(run(testfunc, timeout=30))
```
I added this for `arduino-mega2560` that timeouts when using `test_utils_interactive_sync`, since the test doesn't immediately start. I'll improove the commit or splir.

fjmolinas replied:
Will fix in a different way and split.

fjmolinas replied:
addressed

fjmolinas commented on tests/heap_cmd/tests/01-run.py:
```diff
@@ -12,7 +12,6 @@
 
 def testfunc(child):
     # check startup message
-    child.expect('Shell-based test application for heap functions.')
```
Related, the test starts with the sync command, but this is printed before the shell.

fjmolinas commented on sys/shell/commands/sc_interactive_sync.c:
```diff
@@ -0,0 +1,40 @@
+ */
+
+#include <stdio.h>
+#include "periph/pm.h"
```
indeed.

fjmolinas commented on tests/periph_wdt/Makefile:
```diff
@@ -5,5 +5,6 @@ FEATURES_REQUIRED += periph_wdt
 
 USEMODULE += xtimer
 USEMODULE += shell
+USEMODULE += shell_commands
```
Needed to pull in the sync commands

fjmolinas commented on tests/isr_yield_higher/tests/test.py:
```diff
@@ -11,10 +11,10 @@
-    child.expect('timer triggered')
-    child.expect('first thread done')
-    child.expect('TEST SUCCESSFUL')
+    child.expect_exact('first thread started')
```
removed.

fjmolinas commented:
@aabadie I rebased and took your last comments into account.

aabadie commented on tests/build_system_cflags_spaces/Makefile:
```diff
@@ -1,4 +1,7 @@
 include ../Makefile.tests_common
+
+USEMODULE += test_utils_interactive_sync
```
Is this really needed ? IIUC `test_utils_interactive_sync` is already pulled as default module from line 1.

aabadie commented on tests/thread_msg_seq/main.c:
```diff
@@ -51,9 +53,10 @@ void *sub_thread(void *arg)
-    puts("START");
     msg_t msg;
 
+    test_utils_interactive_sync();
```
You could call this first and keep the use of puts, but after.

fjmolinas commented:
@aabadie addressed, let me know when it is OK to squash and I'll rebase as well.

aabadie commented on tests/README.md:
```diff
@@ -7,6 +7,26 @@ allow basic functionality to be verified as well as provide an example of
 
 
+Implementing automated tests
+-----------------------
```
Minor: the line is not aligned with the text above.

aabadie commented:
I ran compile and test locally on arduino-uno and there are problems:
- some test applications doesn't fit in the small memory of this board anymore:
  ```
  Failures during compilation:
  - [tests/bench_msg_pingpong](tests/bench_msg_pingpong/compilation.failed)
  - [tests/bench_mutex_pingpong](tests/bench_mutex_pingpong/compilation.failed)
  - [tests/bench_thread_flags_pingpong](tests/bench_thread_flags_pingpong/compilation.failed)
  - [tests/bench_thread_yield_pingpong](tests/bench_thread_yield_pingpong/compilation.failed)
  - [tests/driver_ds1307](tests/driver_ds1307/compilation.failed)
  - [tests/msg_try_receive](tests/msg_try_receive/compilation.failed)
  - [tests/pthread_condition_variable](tests/pthread_condition_variable/compilation.failed)
  - [tests/thread_msg_block_w_queue](tests/thread_msg_block_w_queue/compilation.failed)
  - [tests/thread_msg_block_wo_queue](tests/thread_msg_block_wo_queue/compilation.failed)
  ```
- Some tests are failing:
  ```
  - [tests/malloc](tests/malloc/test.failed)
  - [tests/pkg_libfixmath](tests/pkg_libfixmath/test.failed)
  - [tests/trickle](tests/trickle/test.failed)
  ```
  - `tests/malloc` fails because the no allocations are possible on arduino-uno: too little RAM. And then the output is not the one expected. The problem is also present in master. I would decrease the CHUNK_SIZE to fix this test (but is another PR)
  - `tests/pkg_libfixmath`: there's something weird I don't really understand here. It works on master
  -  `tests/trickle`: also doesn't work on master on arduino-uno.

fjmolinas commented:
> some test applications doesn't fit in the small memory of this board anymore:

All of those would be flagged by murdock.

fjmolinas commented:
> tests/pkg_libfixmath: there's something weird I don't really understand here. It works on master

There is an extra `test_interactive_sync` in code that I had removed but was re-added when rebasing because #12687 got merged. Could you re-test? It works for me now.

aabadie commented on tests/msg_avail/main.c:
```diff
@@ -25,18 +25,20 @@
 
     msg_init_queue(msg_queue, MSG_QUEUE_LENGTH);
 
-    puts("[START]");
```
Unrelated, should be readded

fjmolinas replied:
Ok, its a useless print but OK.

aabadie commented:
Please squash

fjmolinas commented:
@aabadie there is some issues with the commits when rebased, let me fix before setting ci

kaspar030 commented on dist/pythonlibs/testrunner/spawn.py:
```diff
@@ -50,6 +51,10 @@ def setup_child(timeout=10, spawnclass=pexpect.spawnu, env=None, logfile=None):
         # make reset yields error on some boards even if successful
         pass
+
+    # Handle synchronization if request by the build system
```
typo: "if requested"

kaspar030 commented on sys/shell/commands/sc_interactive_sync.c:
```diff
@@ -0,0 +1,39 @@
+#include <stdio.h>
+#include "test_utils/interactive_sync.h"
+
+int _s(int argc, char **argv)
```
not sure `_s` or `_r` should be global symbols. IIRC symbols starting with single `_` are supposed to be file scope unique (and thus shouldn't be global).

aabadie replied:
Are you suggesting to change the names all shell command functions in RIOT?
https://github.com/RIOT-OS/RIOT/blob/dd0d5f7fcf90411ca798edcfc839a5a09ddffce8/sys/shell/commands/sc_gnrc_netif.c#L1293

The function names could be a little more explicit: `_test_start`, `_test_ready`.

fjmolinas commented:
Failing tests:

```
    tests/driver_nrfmin/nrf52dk:gnu
    tests/driver_nrfmin/nrf52dk:llvm
    tests/malloc/nrf52dk:llvm
    tests/nordic_softdevice/nrf52dk:gnu
```

The first two and the 4th are fixed by removing the `child.expect_exact("All up, running the shell now")`. I can't reproduce the third failure locally.

aabadie commented:
@fjmolinas, please add `arduino-uno`, `arduino-duemilanove`, `arduino-nano` and `atmega328p` boards to BOARD_INSUFFICIENT_MEMORY and we are good to go with this one!

fjmolinas commented:
I think `tests/periph_wdt` migh have failed because of https://github.com/RIOT-OS/RIOT/pull/12812#issue-345693569.

aabadie commented:
All green!

fjmolinas commented:
@keestux @aabadie Thanks for the review! @MrKevinWeiss Thanks for testing!
